analysis_tools/tracing/slog_trace_subscriber.h

```diff
@@ -0,0 +1,71 @@
+#include <cstring>
+#include <fstream>
+#include <map>
+#include <memory>
+
+#include "slog_cc/util/string_util.h"
+#include "slog_cc/context/subscribers.h"
+#include "slog_cc/context/context.h"
+
+namespace slog {
+
+struct SlogTraceSubscriberState {
+  std::ofstream file;
+  std::map<int64_t, std::string> scope_id_to_name;
+  int64_t min_ts_ns = -1;
+
+  ~SlogTraceSubscriberState() { file << "\n]}\n"; }
+};
+
+struct SlogTraceSubscriber {
+  std::shared_ptr<SlogTraceSubscriberState> state;
+  SlogSubscriber slog_subscriber;
+};
+
+SlogTraceSubscriber CreateSlogTraceSubscriber(
+    const std::string& slog_trace_json_filepath) {
+  auto state = std::make_shared<SlogTraceSubscriberState>();
+
+  state->file.open(slog_trace_json_filepath, std::ios::out | std::ios::trunc);
+  state->file << "{\"traceEvents\": [\n";
+
+  auto json_writer_subscriber =
+      slog::SlogContext::getInstance()->createAsyncSubscriber(
+          [state](const SlogRecord& r) {
+            if (state->min_ts_ns == -1) {
+              state->min_ts_ns = r.time().global_ns;
+            } else {
+              state->file << ",\n";
+            }
```

Comment:
This piece of logic should belong to the state class maybe? That will make the code here a bit more readable maybe